python/ray/data/impl/plan.py

```diff
@@ -348,3 +436,95 @@ def __call__(
         )
         assert isinstance(blocks, BlockList), blocks
         return blocks, stage_info
+
+
+def _rewrite_read_stages(
+    blocks: BlockList,
+    stats: DatasetStats,
+    stages: List[Stage],
+    dataset_uuid: str,
+) -> Tuple[BlockList, DatasetStats, List[Stage]]:
+    """Rewrites read stages into one-to-one stages, if needed."""
+    if _is_lazy(blocks) and stages:
+        blocks, stats, stage = _rewrite_read_stage(blocks)
+        stats.dataset_uuid = dataset_uuid
+        stages.insert(0, stage)
+    return blocks, stats, stages
+
+
+def _rewrite_read_stage(
```

Comment:
Looks like the previous refactoring PR for lazy block list is paying dividend now here.